llvm/lib/Target/AMDGPU/MCTargetDesc/AMDGPUMCExpr.cpp

```diff
@@ -0,0 +1,103 @@
+//===- AMDGPUMCExpr.cpp - AMDGPU specific MC expression classes -----------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "AMDGPUMCExpr.h"
+#include "llvm/MC/MCContext.h"
+#include "llvm/MC/MCStreamer.h"
+#include "llvm/MC/MCSymbol.h"
+#include "llvm/MC/MCValue.h"
+#include "llvm/Support/Allocator.h"
+#include "llvm/Support/raw_ostream.h"
+#include <optional>
+
+using namespace llvm;
+
+const AMDGPUVariadicMCExpr *
+AMDGPUVariadicMCExpr::create(VariadicKind Kind, ArrayRef<const MCExpr *> Args,
+                             MCContext &Ctx) {
+  // Storage for the argument's 'const MCExpr*' allocated through MCContext new
+  // placement which means that AMDGPUVariadicMCExpr objects and all of its
+  // contents will now be allocated through MCContext new placement.
+  //
+  // Will result in an asan failure if allocated on the heap (e.g., through
+  // SmallVector's grow).
+  const MCExpr **CtxArgs = static_cast<const MCExpr **>(
+      Ctx.allocate(sizeof(const MCExpr *) * Args.size()));
+  std::uninitialized_copy(Args.begin(), Args.end(), CtxArgs);
```

Comment:
This should be done in the AMDGPUVariadicMCExpr constructor. Also, I see MCContext does have operator new/delete wrappers around allocate.

Plus you're missing the deallocate which needs to go in the destructor

Reply:
I'm following convention for MCExpr to the extend that I can for AMDGPUVariadicMCExpr: no (target specific) MCExpr is provided the MCContext object and all explicitly have `create` functions to ensure the MCExpr object is allocated through MCContext (i.e., through the MCContext placement new). Additionally, [MCTargetExpr requires subclasses to have trivial destructors as it is never deallocated](https://github.com/llvm/llvm-project/blob/ffe41819e58365dfbe85a22556c0d9d284e746b9/llvm/include/llvm/MC/MCExpr.h#L654-L655). This is further reinforced as MCContext's deallocate function has an [empty body](https://github.com/llvm/llvm-project/blob/ffe41819e58365dfbe85a22556c0d9d284e746b9/llvm/include/llvm/MC/MCContext.h#L851).

Of course, I could be missing something but I do believe this is as good as it gets for supporting variadic arguments in a (target specific) MCExpr.

Reply:
The regular non-target MCExpr's all have an MCContext argument, I don't see why target MCExprs shouldn't have one. It's not clear to me exactly what that comment means by "trivial". I would presume a destructor that calls a deallocate function which does nothing to be trivial. I think it's much better to properly express the alloc/dealloc in the constructor/destructor than relying on whatever the allocation behavior is today

Reply:
> The regular non-target MCExpr's all have an MCContext argument
I believe it's only their `create` functions that have it but it is not forwarded beyond that (i.e., to the MCExpr constructor).

> It's not clear to me exactly what that comment means by "trivial"
I assume it's referring to the C++ standard definition of trivial destructors but I don't believe the MCTargetExpr class itself has a trivial destructor according to those rules.

I'm okay with going off-script with the AMDGPUVariadicMCExpr and adding the MCContext as member + user defined destructor, it doesn't trigger the asan leak either way.